components/locale_canonicalizer/src/locale_canonicalizer.rs

```diff
@@ -184,23 +188,23 @@ fn uts35_check_language_rules(
 
 #[inline]
 fn update_langid(
-    entry: &LanguageIdentifier,
+    entry: &(TinyStr4, TinyStr4, TinyStr4),
     langid: &mut LanguageIdentifier,
 ) -> CanonicalizationResult {
     let mut modified = false;
 
-    if langid.language.is_empty() && !entry.language.is_empty() {
-        langid.language = entry.language;
+    if langid.language.is_empty() && !entry.0.to_string().is_empty() {
+        langid.language = entry.0.parse().unwrap();
```

Comment:
`issue`: I don't think you should be parsing here. I think you should be using https://github.com/unicode-org/icu4x/blob/main/components/locid/src/subtags/language.rs#L122 or introducing a new `From <TinyStr4> for Language` that uses the same thing under the hood.

Reply:
And definitely don't use `entry.0.to_string()`. Instead use https://docs.rs/tinystr/0.4.12/tinystr/struct.TinyStr4.html#method.is_empty